### ISSUE TITLE
Performance Tuning?

### ISSUE THREAD

**Issue by Crisfole** [ractivejs/ractive#1910]
I'm performance issues with larger amounts of data and I'm not really sure how to tune the page.

I was switching an internal sql-running app from Backbone to Ractive and ran a query that returned about 12,000 rows total. It took a *very* long time to render.  My component is here: https://gist.github.com/Crisfole/2e4afdf4c7fcfb387e30. It depends on lodash.

The component gets its `results` set after running a query over a bunch of databases. The format for that object is like this:
```javascript
{
  "database-name": {
    errorMessages: [ "err1", "err2",...],
    messages: ["msg1", "msg2"],
    data: [
        {
          headers: ["column1", "column2", "column3",...],
          rows: [
              ["col1 val", 2, "col 3 val", ...],
              ....
          ]
        }, // one per result set
        ...
    ],
    statistics: {
      "Select-Count": 1,
      "Iud-Count": 2,
      ...
    }
  },
  "database2-name": {
  },
  ...
}
```

There's about 30 databases, each with only one result set containing 200-1000 rows with only 4 columns.

The computed property is misnamed, it should be called 'databaseResultSets' or something like it. It's also very fast, I've checked. The performance in Backbone was very good...but I was creating html inside an underscore template function, so it should be. I'd like to keep it in Ractive, and rendering it w/ _.template inside `{{{` `}}}` feels really icky.

Is there a way to speed this up?

**Comment by Crisfole** [comment]
PS: I realize it's got 4 loops which makes it appx. very bad in terms of  O complexity.

What I can't figure out is why _.template can render the same-ish (uses `for` loops, not {{#each}}) template in under a second...

**Comment by Rich-Harris** [comment]
Initialisation performance in situations like this, where there's a ton of data, is something we can definitely improve. I don't have any ready answers for making this specific example faster, will need to play with it a bit and see if there are any possibilities.

The reason `_.template` is so much quicker is that it's just concatenating strings - the trade-off you get with Ractive is a slower startup in return for faster subsequent updates.

One thing we're looking at is using `fragment.cloneNode(true)` for repeated sections, rather than constructing each fragment inside an `{{#each}}` block programmatically. In theory that should deliver some significant performance boosts for cases like this.

**Comment by fskreuz** [comment]
@Crisfole Wondering if I could play with this one. Can you post the data (or at least something that looks the same in structure and size, if data is sensitive) as the one that is fed into the component?

**Comment by Crisfole** [comment]
The schema I posted above is pretty much it. I'll go see if I can generate some non sensitive data.

**Comment by Rich-Harris** [comment]
Some good news: I generated a bunch of data that roughly matched that schema, and did some profiling - most of the time is spent building up the cache of data that speeds up property lookups. Commenting out [these lines](https://github.com/ractivejs/ractive/blob/1e32783789200d90964cc62c527f7cf57122e5b0/src/viewmodel/prototype/get.js#L89-L96) sped up initialisation by a decent amount (~50%).

A lot of the convoluted caching strategy is rendered totally unnecessary by the work @martypdx is doing on #1740. I checked out that branch, and got a ridiculous performance boost - 12,000 rows went from ~30 seconds to ~2.2. Don't want to set expectations too high - there's still a bunch of failing tests on that branch, and who knows what penalty fixing them will entail - but it bodes pretty well!

**Comment by martypdx** [comment]
> I checked out that branch, and got a ridiculous performance boost - 12,000 rows went from ~30 seconds to ~2.2. Don't want to set expectations too high - there's still a bunch of failing tests on that branch, and who knows what penalty fixing them will entail - but it bodes pretty well!

Wow, that's a nice gain. Expressing the tree structure of the data more naturally in the viewmodel to eliminate those maps was a goal, but I had no idea it would make such a difference in this case.  

The failing tests, while numerous, are in specific known areas that still need to be addressed - object hashes, adaptors, pattern observers and a few of the complex component data sharing situations. So while you're right that there might still be some work entailed that impacts those perf numbers, there's a high chance that for basic data iteration those gains will stick.

**Comment by constantx** [comment]
:+1: sweet! looking forward to the performance boost :)

**Comment by guilhermeaiolfi** [comment]
@Rich-Harris  http://jsperf.com/js-for-loop-vs-array-indexof/268 and this discussion https://github.com/muut/riotjs/issues/484 made me think that a simple indexOf  could actually impact performance in these cases. But I don't think it would be wise to persuit those minor performance gains when #1740 is being worked on. I haven't had any problems regarding performance so far. But it's nice to know it will be even faster in the future.

**Comment by Crisfole** [comment]
Another question. I tried rendering the tables statically, figuring a do it once strategy would speed things up, but it didn't. Any chance we can make square braces skip the expensive caching stuff?

**Comment by Rich-Harris** [comment]
@guilhermeaiolfi that's an interesting result, I wouldn't have expected such a large difference. It'll almost certainly be negligible in the context of #1740, but worth bearing in mind.

@Crisfole If #1740 wasn't in play, I'd say it was worth looking into, but I think it probably makes more sense to focus our efforts on solving this problem at the more fundamental level. If doing things statically is an option, your best bet in the meantime might be to use `_.template` until 0.8, as dirty as that feels. Another option would be rig up some kind of virtual scrolling mechanism, that only shows the rows that are in view. They're kind of a PITA, especially with items of variable height, but they do have an incredible impact on performance with lists of this size.

**Comment by Crisfole** [comment]
@Rich-Harris  yeah, I am already converting that portion to use `_.template`.

I love virtual scrolling, but here I expect my users to print stuff occasionally. I'm very familiar with Slick Grid and its relatives.

**Comment by Rich-Harris** [comment]
Will close this now as it doesn't relate to a specific bug – 0.8 is a decent amount faster than 0.7 across the board, in some cases by huge margins.

**Comment by Crisfole** [comment]
:+1:  thanks all!

**Comment by donnyv** [comment]
So we haven't had an update in a while and I'm having the same issues with over 10,000 records being returned. 

Any news on 0.8?

**Comment by evs-chris** [comment]
There are still a few blocker bugs open on 0.8, mostly dealing with splices, mappings, and observers. Once those are addressed, it's looking like 0.8 can render 10,000 rows of reasonable template in somewhere around 3 seconds on an average-ish machine, half a second on a really good machine, and 5 seconds on a nexus 6p.
